File: src/main/java/com/chimbori/crux/urls/Redirectors.java
package com.chimbori.crux.urls;

import com.chimbori.crux.common.StringUtils;

import java.net.URI;
import java.net.URISyntaxException;

class Redirectors {
  private Redirectors() {
  }

  /**
   * Defines a pattern used by a specific service for URL redirection. This should be stateless,
   * and will be called for each URL that needs to be resolved.
   */
  abstract static class RedirectPattern {
    /**
     * @return true if this RedirectPattern can handle the provided URL, false if not.
     */
    public abstract boolean matches(URI uri);

    /**
     * @return the actual URL that is pointed to by this redirector URL.
     */
    public abstract URI resolve(URI uri) throws URISyntaxException;

    /**
     * To avoid every implementing class wrap their own code with a try-catch, they are encouraged
     * to use the resolve() method which throws {@link URISyntaxException}, while callers can
     * use this method which simply wraps that method’s return value in a try-catch.
     */
    URI resolveHandlingException(URI uri) {
      try {
        return resolve(uri);
      } catch (URISyntaxException e) {
        return uri;
      }
    }
  }

  static final RedirectPattern[] REDIRECT_PATTERNS = {
      new RedirectPattern() {  // Facebook.
        @Override
        public boolean matches(URI uri) {
          return uri.getHost() != null
              && uri.getHost().endsWith(".facebook.com")
              && "/l.php".equals(uri.getPath());
        }

        @Override
        public URI resolve(URI uri) throws URISyntaxException {
          return new URI(StringUtils.getQueryParameters(uri).get("u"));
        }
      },
      new RedirectPattern() {  // Google.
        @Override
        public boolean matches(URI uri) {
          return uri.getHost() != null
              && uri.getHost().endsWith(".google.com")
              && "/url".equals(uri.getPath());
        }

        @Override
        public URI resolve(URI uri) throws URISyntaxException {
          return new URI(StringUtils.getQueryParameters(uri).get("q"));
        }
      }
  };
}

File: src/main/java/com/chimbori/crux/urls/Redirectors.kt
package com.chimbori.crux.urls

import com.chimbori.crux.common.StringUtils.getQueryParameters
import java.net.URI

internal object Redirectors {
  val REDIRECT_PATTERNS = listOf(
      object : RedirectPattern {  // Facebook.
        override fun matches(uri: URI) = uri.host?.endsWith(".facebook.com") == true && uri.path == "/l.php"
        override fun resolve(uri: URI) = URI(getQueryParameters(uri)["u"])
      },
      object : RedirectPattern { // Google.
        override fun matches(uri: URI) = uri.host?.endsWith(".google.com") == true && uri.path == "/url"
        override fun resolve(uri: URI) = URI(getQueryParameters(uri)["q"])
      }
  )

  /**
   * Defines a pattern used by a specific service for URL redirection. This should be stateless, and will be called for
   * each URL that needs to be resolved.
   */
  internal interface RedirectPattern {
    /** @return true if this RedirectPattern can handle the provided URL, false if not. */
    fun matches(uri: URI): Boolean

    /** @return the actual URL that is pointed to by this redirector URL. */
    fun resolve(uri: URI): URI
  }
}
